### DISCUSSION TITLE
How does type checking work in the VSCode Extension for MDX with experimentalLanguageServer enabled?

### DISCUSSION THREAD

**Discussion by thomfoolery** [mdx-js/mdx#2396] [Q&A] [answered]
We're trying to improve the content authoring experience for documentation written in MDX.

The `eslint-plugin-mdx` has helped a lot, but we expected the [VS Code Extension for MDX](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx) to type check our component props, and other IntelliSense features like autocomplete. This would really help our authors understand what might be wrong when they are writing interactive tutorials.

We have enabled `mdx.experimentalLanguageServer` but we're not seeing any IntelliSense features. We're questioning if maybe we haven't configured or setup our project properly, but are struggling to figure out how/where to start debugging.

We imagine the `mdx-analyzer` is not aware of where the source code is for the components we provide through a MDXProvider and therefor can not provide type checking or introspection abilities.

Is there away providing additional information to the extension/language server to provide better IntelliSense assistance to our authors?

Here are the relevant issues I've uncovered
- MDX +TypeScript support ([github discussion](https://github.com/orgs/mdx-js/discussions/1562))
- Typescript - support for MDX ([github issue](https://github.com/microsoft/TypeScript/issues/36440))
- ESLint MDX - type check React component props ([github issue](https://github.com/mdx-js/eslint-mdx/issues/464))
- MDX Analyzer - Support components ([github issue](https://github.com/mdx-js/mdx-analyzer/issues/260))

**Comment by remcohaszing** [comment, marked as answer]
I’ve just released [version 1.5.0](https://github.com/mdx-js/mdx-analyzer/releases/tag/vscode-mdx%401.5.0) of the MDX extension. Its IntelliSense got a complete rewrite, but I think you’ll run into the same issue. IntelliSense works, but it’s subtle. I think you want https://github.com/mdx-js/mdx-analyzer/issues/352 to be implemented.

**Reply by thomfoolery** [reply]
Thank you @remcohaszing for the prompt reply and your hard work on improving the MDX authorship experience!

I understand there's still work to be done and we'd love to help out, 
but we have no idea where to start to address your [issue](https://github.com/mdx-js/mdx-analyzer/issues/352).

Also, 
We're curious to understand if we would need to build a mechanism (config file?) to make the `mdx-analyzer` aware of components provided through a `MDXProvider`?

**Reply by karlhorky** [reply]
Would also love if full type checking for the components provided through `MDXProvider` was in the toolchain somewhere.

Our use case: checking whether a value of a particular slug is valid and failing CI if invalid (currently set to a union of string literals in TS):

```mdx
- 🎥 slides: <LectureSlideDeck slug="js-fundamentals" />
```

Cross-linking to the `eslint-plugin-mdx` issue, in case this ends up with progress:

- https://github.com/mdx-js/eslint-mdx/issues/464

**Reply by karlhorky** [reply]
@remcohaszing or maybe after the `checkMdx` option from #387 checking against `MDXProvider` components is still not possible yet?

If so, what would be your recommendations for steps how I can write my own small Node.js CLI to check the types of props in an `.mdx` document on GitHub Actions? (to be clear, props of components provided by the `MDXProvider`, not components `import`ed statically at the top of the `.mdx` document)

Eg. my own simple (and maybe naive) first version of what you're mentioning here:

- https://twitter.com/remcohaszing/status/1755867542457897271

Or do you think it's faster for me to pursue a small `eslint-plugin-mdx` MVP with @JounQin?

- https://github.com/mdx-js/eslint-mdx/issues/464

**Reply by JounQin** [reply]
I don't have much time to work on this recently, but I have a wish that we can write TypeScript codes in `.mdx` directly without jsdoc.

**Reply by karlhorky** [reply]
> I have a wish that we can write TypeScript codes in `.mdx` directly without jsdoc.

That makes sense - I think that would be a common ask from the community. JSDoc is nice as an alternative, but has non-insignificant downsides: verbose, harder to use, less common (harder to find resources for), doesn't have all features of TS without extra `.d.ts` files.

---

That's not what I and @thomfoolery are currently looking for though - as mentioned above, I'm looking for a way to type check usages of the React components that are passed to MDX via the `MDXProvider`.

As @thomfoolery also guessed, it is possibly a limitation of `mdx-analyzer` that this doesn't work:

> We imagine the `mdx-analyzer` is not aware of where the source code is for the components we provide through a MDXProvider and therefor can not provide type checking or introspection abilities

**Reply by remcohaszing** [reply]
Support for type checking the `components` prop is tracked in https://github.com/mdx-js/mdx-analyzer/issues/260 and is almost done. At first I thought this was going to include support for the MDX provider, but in the end it won’t. I created a new issue for the MDX provider https://github.com/mdx-js/mdx-analyzer/issues/399.

**Reply by remcohaszing** [reply]
The reason there’s not CLI yet for MDX type checking is because:

- I haven’t looked into it yet.
- I want to stable editor support before making CI fail for people.
- I want this to interoperate with other Volar based language tooling. I imagine a single Volar CLI would be useful for type checking a project that combines Astro, Emver, MDX, Svelte, Vue, and/or other frameworks that might leverage Volar in the future.

**Reply by karlhorky** [reply]
Totally makes sense, thanks for your work here and looking forward to your implementation!

If I was to try to get a small, naive MVP Node.js CLI for a type checker for `.mdx` for myself (not covering edge cases), do you have a suggestion for how to go about this?

Would you suggest something like the following? (maybe I'm missing a step)

1. Extract `components` object from inside `MDXProvider` prop to export
2. Compile MDX using [`@mdx-js/mdx`](https://mdxjs.com/packages/mdx/)
3. Run type checking using `tsc` on the compiled file(s) (maybe these are `VFile`s?)
4. Delete the compiled file(s)

(also open to options which are more hacky / require fewer steps, eg. [augmenting `@types/mdx`](https://github.com/mdx-js/mdx-analyzer/issues/399#:~:text=augmenting%20%40types/mdx))

**Reply by remcohaszing** [reply]
I really strongly recommend to base that on Volar. I know there’s code for doing something like that. There’s already `vue-tsc`, maybe something for Astro too. I just haven’t gotten to it for MDX yet.

**Comment by karlhorky** [comment]
For anyone looking for the current research on CLI TypeScript type checking in `*.mdx` files, some current research and first implementation experiments can be found here:

- https://github.com/mdx-js/mdx-analyzer/issues/292
- https://github.com/volarjs/volar.js/issues/145

Keywords: command line, CLI, terminal, Volar.js, MDX Analyzer
